src/Microsoft.Health.Fhir.CosmosDb/Features/Storage/CosmosDbCollectionPhysicalPartitionInfo.cs

```diff
@@ -0,0 +1,166 @@
+                }
+                catch (OperationCanceledException) when (cancellationToken.IsCancellationRequested)
+                {
+                    return;
+                }
+                catch (Exception e)
+                {
+                    _logger.LogError("Unable to get physical partition count.", e);
+                }
+            }
+        }
+
+        private async Task<int> GetPhysicalPartitionCount(CancellationToken cancellationToken)
+        {
+            using HttpClient client = _httpClientFactory.CreateClient();
+
+            string host = _dataStoreConfiguration.Host;
+            string key = _dataStoreConfiguration.Key;
+
+            if (string.IsNullOrWhiteSpace(host) && string.IsNullOrWhiteSpace(key))
+            {
+                host = CosmosDbLocalEmulator.Host;
+                key = CosmosDbLocalEmulator.Key;
+            }
+
+            string date = DateTime.UtcNow.ToString("R");
+
+            bool isResourceToken = IsResourceToken(key);
+
+            string authToken = HttpUtility.UrlEncode(
+                isResourceToken
+                    ? key
+                    : GenerateAuthToken(
+                        "get",
+                        "pkranges",
+                        $"dbs/{_dataStoreConfiguration.DatabaseId}/colls/{_collectionConfiguration.CollectionId}",
+                        date,
+                        key));
+
+            using var httpRequestMessage = new HttpRequestMessage(
+                HttpMethod.Get,
+                $"{host}/dbs/{_dataStoreConfiguration.DatabaseId}/colls/{_collectionConfiguration.CollectionId}/pkranges")
+            {
+                Headers =
+                {
+                    { "authorization", authToken },
+                    { "x-ms-version", "2018-12-31" },
+                    { "x-ms-date", date },
+                },
+            };
+
+            HttpResponseMessage httpResponseMessage = await client.SendAsync(httpRequestMessage, cancellationToken);
+            if (httpResponseMessage.IsSuccessStatusCode)
+            {
+                var partitionKeyRangesResponse = await httpResponseMessage.Content.ReadFromJsonAsync<PartitionKeyRangesResponse>(new JsonSerializerOptions { PropertyNameCaseInsensitive = true }, cancellationToken);
+                return partitionKeyRangesResponse.PartitionKeyRanges.Length;
+            }
+
+            if (httpResponseMessage.StatusCode == HttpStatusCode.TooManyRequests)
+            {
+                throw new RequestRateExceededException(null);
+            }
+
+            _logger.LogError("Unable to get physical partition count. {ResponseBody}", await httpResponseMessage.Content.ReadAsStringAsync(cancellationToken));
+
+            httpResponseMessage.EnsureSuccessStatusCode();
+            return 0; // will not reach this
+        }
+
+        public static bool IsResourceToken(string key) => key.StartsWith("type=resource&", StringComparison.InvariantCulture);
```

Comment:
Does this need a nullcheck?

Reply:
Made this method private and added validation for key and host in the calling method.